src/compressor/QatAccel.cc

```diff
@@ -48,22 +53,111 @@ bool QatAccel::init(const std::string &alg) {
   rc = qzSetDefaults(&params);
   if (rc != QZ_OK)
       return false;
+  return true;
+}
 
+static bool SetupSession(QatAccel::session_ptr &session, QzSessionParams_T &params) {
+  int rc;
   rc = qzInit(session.get(), QZ_SW_BACKUP_DEFAULT);
-  if (rc != QZ_OK && rc != QZ_DUPLICATE && rc != QZ_NO_HW)
+  if (rc != QZ_OK && rc != QZ_DUPLICATE)
     return false;
-
   rc = qzSetupSession(session.get(), &params);
-  if (rc != QZ_OK && rc != QZ_DUPLICATE && rc != QZ_NO_HW ) {
-    qzTeardownSession(session.get());
-    qzClose(session.get());
+  if (rc != QZ_OK) {
+    return false;
+  }
+  return true;
+}
+
+// put the session back to the session pool in a RAII manner
+struct cached_session_t {
+  cached_session_t(QatAccel* accel)
+    : accel{accel} {
+    std::unique_lock lock{accel->mutex};
+    if (!accel->sessions.empty()) {
+      session = std::move(accel->sessions.back());
+      accel->sessions.pop_back();
+      return;
+    }
+
+    // If there are no available session to use, we try allocate a new
+    // session. When we failed to allocate a new session, we block on here,
+    // waiting a available session;
+    QzSessionParams_T params = {(QzHuffmanHdr_T)0,};
+    QatAccel::session_ptr s{new struct QzSession_S()};
+    memset(s.get(), 0, sizeof(struct QzSession_S));
+    if (getQzParams(accel->alg, params) && SetupSession(s, params)) {
+      session = std::move(s);
+    } else {
+      accel->cond.wait(lock, [accel] { return !accel->sessions.empty(); });
+      session = std::move(accel->sessions.back());
+      accel->sessions.pop_back();
+    }
+  }
+  ~cached_session_t() {
+    std::scoped_lock lock{accel->mutex};
+    // if sessions.size() is large and equal than qat_compressor_session_max_number, we release current session.
+    uint64_t sessions_num = g_ceph_context->_conf.get_val<uint64_t>("qat_compressor_session_max_number");
+    if (accel->sessions.size() < sessions_num) {
+      accel->sessions.push_back(std::move(session));
+      accel->cond.notify_one();
+    }
+  }
+  struct QzSession_S* get() {
+    assert(static_cast<bool>(session));
+    return session.get();
+  }
+  QatAccel* accel;
+  QatAccel::session_ptr session;
+};
+
+QatAccel::QatAccel() {}
+
+QatAccel::~QatAccel() {
+  // First, we should uninitialize all QATzip session that disconnects all session
+  // from a hardware instance and deallocates buffers.
+  sessions.clear();
+  // Then we close the connection with QAT.
+  // There 1 is meaningless, only make sure the para is no nullptr.
+  qzClose((QzSession_T*)1);
+}
+
+bool QatAccel::init(const std::string &alg) {
+  QzSessionParams_T params = {(QzHuffmanHdr_T)0,};
+  std::scoped_lock lock(mutex);
+  if (init_called) {
+    return true;
+  } else {
```

Comment:
nit, just put
```c++
if (init_called) {
  return true;
}

// generate the sessions
```

would be better in the sense of readability. as the if clauses are not symmetric at all: one `returns` while the other does not.

Reply:
done